pkg/libinit/root_linux.go

```diff
@@ -263,24 +275,97 @@ func InstallModules(pattern string, exclude map[string]bool) error {
 	for _, filename := range files {
 		f, err := os.Open(filename)
 		if err != nil {
-			log.Printf("installModules: can't open %q: %v", filename, err)
+			log.Printf("InstallModules: can't open %q: %v", filename, err)
 			continue
 		}
-		// Module flags are passed to the command line in the form modulename.flag=val
+		// Module flags are passed to the command line in the from modulename.flag=val
 		// And must be passed to FileInit as flag=val to be installed properly
 		moduleName := strings.TrimSuffix(filepath.Base(filename), filepath.Ext(filename))
 		if _, ok := exclude[moduleName]; ok {
-			log.Printf("Skipping module %s", moduleName)
+			log.Printf("Skipping module %q", moduleName)
 			continue
 		}
 
 		flags := cmdline.FlagsForModule(moduleName)
-		err = kmodule.FileInit(f, flags, 0)
+		if err = kmodule.FileInit(f, flags, 0); err != nil {
+			log.Printf("InstallModules: can't install %q: %v", filename, err)
+		}
 		f.Close()
+	}
+
+	return nil
+}
+
+func readModules(f *os.File) []string {
+	r := bufio.NewReader(f)
+	modules := []string{}
+	for {
+		if i, err := r.ReadString('\n'); err == nil {
+			i = strings.TrimSpace(i)
+			if i == "" {
+				continue
+			} else if strings.HasPrefix(i, "#") {
+				continue
+			} else {
+				modules = append(modules, strings.TrimSuffix(i, "\n"))
+			}
+		} else if err != io.EOF {
+			log.Println("error on reading:", err)
+		} else {
+			break
+		}
+	}
+	return modules
+}
+
+// InstallModules installs kernel modules from .conf files in /etc/modules-load.d/
+func InstallModulesFromModulesLoad(pattern string, exclude map[string]bool) error {
+	files, err := filepath.Glob(pattern)
+	if err != nil {
+		return err
+	}
+
+	for _, filename := range files {
+		f, err := os.Open(filename)
 		if err != nil {
-			log.Printf("installModules: can't install %q: %v", filename, err)
+			log.Printf("InstallModulesFromModulesLoad: can't open %q: %v", filename, err)
+			continue
+		}
+		defer f.Close()
+		fModules := readModules(f)
+		for _, moduleName := range fModules {
+			if _, ok := exclude[moduleName]; ok {
+				log.Printf("Skipping module %q", moduleName)
+				continue
+			}
+			flags := cmdline.FlagsForModule(moduleName)
+			err = kmodule.Probe(moduleName, flags)
+			if err != nil {
+				log.Printf("InstallModulesFromModulesLoad: can't install %q: %v", moduleName, err)
+			}
 		}
 	}
+	return nil
+}
 
+// InstallModules installs kernel modules from .conf files in /etc/modules-load.d/
+func InstallModulesFromCmdline(exclude map[string]bool) error {
+	modules, present := cmdline.Flag("modules_load")
```

Comment:
Right here is where you can use the new cmdline functions to create fake command lines and parse them.

Reply:
OK, what I'd suggest is that these functions change in a few ways. We might want a struct containing needed information, and one item in that struct could be a *CmdLine created from the cmdline package. That CmdLine can then be mocked up for testing or derived from the real cmdline. This should let us right much better tests.

Reply:
I'll take a few stabs at this tomorrow. I'm not quite happy with this code as-is so I won't mind do a bit of reworking :)

Reply:
how goes? I'd like to get this in.